### PR TITLE
Move find-refs from object browser over to the new streaming-far-window.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
retest perf_correctness_prtest  please

**Comment by sharwell** [comment]
Blocked on #18840.

**Comment by CyrusNajmabadi** [comment]
That PR has been merged in now.  Please rereview.  Thanks!

File: src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs [hunk 1]
```diff
@@ -508,5 +512,51 @@ protected override bool TryExec(Guid commandGroup, uint commandId)
...
+            {
+                return _streamingPresenters.FirstOrDefault()?.Value;
+            }
+            catch
+            {
+                return null;
+            }
+        }
+
+        private async void FindReferences(
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs]: ❕  Missing `Async` suffix.
❕  Don't use `async void`. Return a `Task` even if you don't intend to wait on it specifically.

File: src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs [hunk 2]
```diff
@@ -508,5 +512,51 @@ protected override bool TryExec(Guid commandGroup, uint commandId)
...
             return false;
         }
+
+        private IStreamingFindUsagesPresenter GetStreamingPresenter()
+        {
+            try
+            {
+                return _streamingPresenters.FirstOrDefault()?.Value;
+            }
+            catch
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs]: ❔ What are we expecting this to throw? Will it be logged anywhere?

File: src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs [hunk 3]
```diff
@@ -508,5 +512,51 @@ protected override bool TryExec(Guid commandGroup, uint commandId)
...
+                var context = presenter.StartSearch(
+                    EditorFeaturesResources.Find_References, supportsReferences: true);
+
+                var cancellationToken = context.CancellationToken;
+
+                // Fire and forget the work to go get references.
+                var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                var symbol = symbolListItem.ResolveSymbol(compilation);
+
+                await AbstractFindUsagesService.FindSymbolReferencesAsync(
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs]: :memo: `FindSymbolReferencesAsync` will throw a `NullReferenceException` if `symbol` is null.

File: src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs]: > Return a Task even if you don't intend to wait on it specifically.

Sure.

File: src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs]: good catch!

File: src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs]: done.

File: src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs]: done

File: src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs]: This shouldn't be needed.  It was a holdover to the days where our code and the editor's code for FAR were moving around the divisoin through different RIs/FIs.  Now that everything has flowed everywhere, i can remove this.

File: src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs [hunk 4]
```diff
@@ -501,7 +501,12 @@ protected override bool TryExec(Guid commandGroup, uint commandId)
                             var project = this.Workspace.CurrentSolution.GetProject(symbolListItem.ProjectId);
                             if (project != null)
                             {
-                                FindReferences(streamingPresenter, symbolListItem, project);
+                                // Note: we kick of FindReferencesAsync in a 'fire and forget' manner.
+                                // We don't want to block the UI thread while we compute the references,
+                                // and the references will be asynchronously added to the FindReferences
+                                // window as they are computed.  The user also knows something is happening
+                                // as the window, with the progress-banner will pop up immediately.
+                                var task = FindReferencesAsync(streamingPresenter, symbolListItem, project);
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs]: :question: Does this need a `Task.Run` to ensure the operation really runs in the background?

Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractObjectBrowserLibraryManager.cs]: I can do that.

**Comment by CyrusNajmabadi** [comment]
Change made.  Plx rereview when you get a chance.

**Comment by CyrusNajmabadi** [comment]
Tagging @MattGertz 

I am currently working on a large set of refactorings to FindReferences to make it so that we can move it *entirely* out of proc.  By *entirely* i mean that all the expensive computational work will be done in the OOP process, and only 'simple data results' will be passed back to VS to only be shown in the presentation layer.  

Right now, the approach we take is one where we can do *much* of the expensive computation OOP.  HOwever, when we pass the results back to VS we may still encur a lot of expensive computation.  SPecifically, we may need to go back and create compilations (expensive) to resolve symbols we pass back.  And we end up needing to go back compilations/semantic-models (expensive) to get the classified txt we display in the UI.

This work is part of a long sequence of refactorings that addresses all this.  I anticipate a handful of additional PRs until this done (ideally next week).  But i'm trying to get independent pieces in to prevent hte overall change+PRs from being too large.

**Comment by MattGertz** [comment]
@CyrusNajmabadi Sounds good, I look forward to seeing the results.  For the upcoming ones related to this particular FAR re-work, we can treat it as TELL mode -- tag me, but don't block on me.  When it's all done, I'd love to see some comparative numbers.

**Comment by CyrusNajmabadi** [comment]
> When it's all done, I'd love to see some comparative numbers.

Sure.  I do expect some potential improvements in FAR and NavigateTo numbers (generally by being able to parallize things).  However, that's not the primary motivation here.  The core goal here is to remove expensive, heavy-allocation, global features out of VS.   That way we place a much lower burden on the memory-system/GC in VS, which can translate out to a leaner VS environment which needs many less pauses that might interrupt the user.  

So, this is really about removing the types of patterns that end up causing intermittent latency spikes, while not really being intended to change much about how the specific features themselves work.

**Pull request merged**